cmd/multiclusterapp.go

```diff
@@ -0,0 +1,792 @@
+				Usage:       "List templates available for installation",
+				Description: "\nList all app templates in the current Rancher server",
+				ArgsUsage:   "None",
+				Action:      globalTemplateLs,
+				Flags: []cli.Flag{
+					formatFlag,
+					cli.StringFlag{
+						Name:  "catalog",
+						Usage: "Specify the catalog to list templates for",
+					},
+				},
+			},
+			cli.Command{
+				Name:        "show-template",
+				Aliases:     []string{"st"},
+				Usage:       "Show versions available to install for an app template",
+				Description: "\nShow all available versions of an app template",
+				ArgsUsage:   "[TEMPLATE_ID]",
+				Action:      templateShow,
+			},
+			cli.Command{
+				Name:      "show-app",
+				Aliases:   []string{"sa"},
+				Usage:     "Show an app's available versions and revisions",
+				ArgsUsage: "[APP_NAME/APP_ID]",
+				Action:    showMultiClusterApp,
+				Flags: []cli.Flag{
+					formatFlag,
+				},
+			},
+		},
+	}
+}
+
+func multiClusterAppLs(ctx *cli.Context) error {
+	c, err := GetClient(ctx)
+	if err != nil {
+		return err
+	}
+
+	collection, err := c.ManagementClient.MultiClusterApp.List(defaultListOpts(ctx))
+	writer := NewTableWriter([][]string{
+		{"NAME", "App.Name"},
+		{"STATE", "App.State"},
+		{"VERSION", "Version"},
+		{"TARGET_PROJECTS", "Targets"},
+	}, ctx)
+
+	defer writer.Close()
+
+	templateVersionCache := map[string]*managementClient.TemplateVersion{}
+	clusterCache := map[string]*managementClient.Cluster{}
+	projectCache := map[string]*managementClient.Project{}
+	for _, item := range collection.Data {
+		version, err := getTemplateVersion(c.ManagementClient, templateVersionCache, item.TemplateVersionID)
+		if err != nil {
+			return err
+		}
+		targetNames, err := getReadableTargetNames(c, clusterCache, projectCache, item.Targets)
+		if err != nil {
+			return err
+		}
+		writer.Write(&MultiClusterAppData{
+			App:     item,
+			Version: version,
+			Targets: strings.Join(targetNames, ","),
+		})
+	}
+	return writer.Err()
+}
+
+func getTemplateVersion(client *managementClient.Client, templateVersionCache map[string]*managementClient.TemplateVersion, ID string) (string, error) {
+	var version string
+	if tv, ok := templateVersionCache[ID]; ok {
+		version = tv.Version
+	} else {
+		templateVersion, err := client.TemplateVersion.ByID(ID)
+		if err != nil {
+			return "", err
+		}
+		templateVersionCache[templateVersion.ID] = templateVersion
+		version = templateVersion.Version
+	}
+	return version, nil
+}
+
+func getReadableTargetNames(c *cliclient.MasterClient, clusterCache map[string]*managementClient.Cluster, projectCache map[string]*managementClient.Project, targets []managementClient.Target) ([]string, error) {
+	var targetNames []string
+	var err error
+	for _, target := range targets {
+		projectID := target.ProjectID
+		clusterID, _ := parseScope(projectID)
+		cluster, ok := clusterCache[clusterID]
+		if !ok {
+			cluster, err = c.ManagementClient.Cluster.ByID(clusterID)
```

Comment:
This needs to go back to the list but also include pagination, [here](https://github.com/superseb/cli/blob/3b590c5dc4e6c369bd4119aaa576c4c6e9a8cb3a/cmd/catalog.go#L455) is an example of what that looks like. The list is going to be more efficient so that should be used for the cluster and projects.

Reply:
Updated to use cache in paginated lists.